includes/admin/class-amp-cache-utilities.php

```diff
@@ -0,0 +1,251 @@
+		$amp_cache_resource_path = self::get_amp_cache_path_for_url( $url, $content_type );
+		if ( $amp_cache_resource_path ) {
+			return self::$amp_cache_update_url_base . '/' . ltrim( $amp_cache_resource_path, '/' );
+		}
+		return false;
+	}
+
+	/**
+	 * Given a post_id or WP_Post onbject, calculate the AMP cache path part (the part after 
+	 * https://cdn.ampproject.org) for this post.
+	 * 
+	 * @param  int/WP_Post	$post 			The post to get rhe AMP cache path for.
+	 * @param  string 		$content_type 	Currently only supports 'c' for posts and 'i' for images.
+	 *                                 		Optional, if provided, this overrides the actual content
+	 *                                 		type of the post. If null, the content type is determined
+	 *                                 		by post_type.
+	 * @param  string 	$scheme       		Optional. 'http' or 'https' If provided, this overrides the
+	 *                                		scheme in the URL. If null, scheme is taken from the URL.
+	 * @return string/bool              	The AMP cache path part on success, false on failure.
+	 */
+	public static function get_amp_cache_path_for_post( $post, $content_type = null, $scheme = null ) {
+		$permalink = get_permalink( $post );
+
+		// If permalink couldn't be retrieved, return failure.
+		if ( false === $permalink ) {
+			return false;
+		}
+
+		// determine $content_type, if not specified
+		if ( null == $content_type ) {
+			$post_type = get_post_type( $post );
+			switch ( $post_type ) {
+				case 'attachment':
+					$content_type = 'i';
+					break;
+				case 'post':
+					$content_type = 'c';
+					break;
+				default:
+					// unhandled post type
+					return false;
+			}
+		}
+
+		// Return the url
+		return self::get_amp_cache_path_for_url( $permalink, $content_type, $scheme );
+	}
+
+	/**
+	 * Given a URL and content_type, calculate the AMP cache path part (the part after
+	 * https://cdn.ampproject.org) for this url/content_type
+	 * .
+	 * @param  string 	$url     		The url to calculate the AMP cache path from.
+	 * @param  string 	$content_type 	Currently only supports 'c' for posts and 'i' for images.
+	 * @param  string 	$scheme       	Optional. 'http' or 'https' If provided, this overrides the
+	 *                                	scheme in the URL. If null, scheme is taken from the URL.
+	 * @return string/bool              The AMP cache path part on success, false on failure.
+	 */
+	public static function get_amp_cache_path_for_url( $url, $content_type , $scheme = null ) {
+		$parsed_url = wp_parse_url( $url );
+		// If permalink couldn't be parsed, then return failure.
+		if ( false === $parsed_url ) {
+			return false;
+		}
+
+		// If there is no host part to this URL, return failure.
+		if ( ! isset( $parsed_url['host'] ) ) {
+			return false;
+		}
+
+		// If there is no scheme specified in the parameter list and this is a protocol
+		// relative URL, then we can't figure out whether this should be https or http.
```

Comment:
The AMP Cache doesn't treat http and https the same.  For example, if you ping https://cdn.ampproject.org/update-ping/c/s/example.com, but the URL is actually http://example.com, then nothing gets cached.

So, I suppose that another option is to try both .../c/s/example.com and .../c/example.com instead of failing.